### PR TITLE
chore: `dedent`

### PR DESCRIPTION
**Pull request by Rich-Harris**
Per https://github.com/sveltejs/kit/pull/9270#issuecomment-1450855548, this adds a `dedent` utility for the places where we need to generate code from template literals. It allows our code to be much more readable, and makes the generated code more reliably readable as well.

Need to use it for the virtual modules declared inline, and figure out whether to use it in `render.js` (I think the answer is probably 'yes', but want to make sure we can do so without introducing any performance overhead)

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [ ] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [ ] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

**Comment by Rich-Harris** [comment]
I think the `render.js` stuff should probably happen as part of a separate refactor, if at all

File: packages/kit/src/core/sync/utils.js [hunk 1]
```diff
@@ -25,9 +25,44 @@ export function write(file, code) {
...
-/** @param {string} str */
-export function trim(str) {
-	const indentation = /** @type {RegExpExecArray} */ (/\n?([ \t]*)/.exec(str))[1];
-	const pattern = new RegExp(`^${indentation}`, 'gm');
-	return str.replace(pattern, '').trim();
+/** @type {WeakMap<TemplateStringsArray, { strings: string[], indents: string[] }>} */
+const dedent_map = new WeakMap();
+
+/**
+ * @param {TemplateStringsArray} strings
```
Comment (benmccann) [inline packages/kit/src/core/sync/utils.js:32]: a comment here could help. is this the opposite of indent? why do we need it? won't Vite handle minimization for us? it seems like outputting minimized code in the intermediate step might make debugging harder

Comment (Rich-Harris) [reply packages/kit/src/core/sync/utils.js:33]: added a comment. it's just a dedenting function. it takes a string like 

```js
			const code = dedent`
				const things = [
					${things.join('\n')}
				];
			`;
```

and makes it as though you wrote this

```js
			const code = `const things = [\n\t${things.join('\n\t')}\n];`
```

File: packages/kit/src/core/sync/utils.js [hunk 2]
```diff
@@ -29,6 +29,7 @@ export function write(file, code) {
 const dedent_map = new WeakMap();
 
 /**
+ * Removes leading whitespace from each line of the input string
```
Comment (benmccann) [inline packages/kit/src/core/sync/utils.js:32]: If I understand correctly now, it doesn't remove all leading whitespace, but just some of it
```suggestion
 * Allows indenting template strings without the extra indentation ending up in the result.
```

**Review by benmccann — APPROVED** [review]

**Pull request merged**